### ISSUE TITLE
Support custom junction table names

### ISSUE THREAD

**Issue by devinivy** [balderdashy/waterline-schema#20]
Should be an attribute property set on the dominant side of a many-to-many association.  See discussion https://github.com/balderdashy/waterline/issues/725 .  Relevent question on #15 .

**Comment by dmarcelino** [comment]
Thanks for raising this, this is really an important feature for graphDBs where junction tables are usually named as verbs.

**Comment by kirkaleze** [comment]
Are custom Join tables supported yet? I need to create a custom Join table that I can add additional attributes to. 

When I use the 'through' attribute, I get the following error in the console:
 
    this.collections[linkedCollection].junctionTable = true;
                                                     ^
    TypeError: Cannot set property 'junctionTable' of undefined

I'm using sails v0.11 and Waterline-schema is at 0.1.17. 

The example I am working with is Persons and Beers. I would like to create a many-to-many relationship to query all the beers a person has tried, as well as all of the persons that have tried a particular beer. I've created a join table and model with custom attributes for drankOn (datetime) and rating(int). 

##### Person.js
```
  attributes: {
    personID: {type: 'int', primaryKey: true, unique: true, autoIncrement: true, notNull: true},
    personName: {type: 'string', notNull: true},

    beers: {
      collection: 'beer',
      via: 'persons',
      through: 'PersonBeer'
    },
  }
```


##### Beer.js
```
   attributes: {
    beerID: {type: 'int', primaryKey: true, unique: true, autoIncrement: true, notNull: true},
    beerName: {type: 'string', notNull: true},

    persons: {
      collection: 'person',
      via: 'beers',
      through: 'PersonBeer'
    },
  }
```
##### PersonBeer.js
```
  tableName: 'personbeer',
  tables: ['person', 'beer'],
  junctionTable: true,

  attributes: {
    personBeerID: {type: 'int', primaryKey: true, unique: true, autoIncrement: true, notNull: true},

    person_beers: {
      columnName: 'beerID',
      type: 'integer',
      foreignKey: true,
      references: 'person',
      on: 'personID',
      via: 'beer_persons',
      groupBy: 'person'
    },

    beer_persons: {
      columnName: 'personID',
      type: 'integer',
      foreignKey: true,
      references: 'beer',
      on: 'beerID',
      via: 'person_beers',
      groupBy: 'beer'
    },

    drankOn: {type: 'datetime', defaultsTo: function () {return new Date();}, notNull: true},
    rating: {type: 'int'}
  }
```

Any ideas on what we can do to get this to work?

**Comment by devinivy** [comment]
Perhaps try lowercasing your `through` attributes?  Bear in mind that for now through tables are "working" but not officially supported with documentation.

**Comment by kirkaleze** [comment]
Ok, I was actually able to lift Sails after making that change. Now I can try to work with some data. Thank you!

**Comment by kirkaleze** [comment]
When Using:

```
Person.findOne(1).exec(function(err, person) {
	person.beers.add(2);
	person.save(function(err){});
});
```

I am getting:

```
TypeError: Cannot read property 'via' of undefined
```

I know it's not documented or supported yet, but does 'Through' actually work in any form yet? Is there any variation of the above statement that will allow me to use a Through statement?

**Comment by dmarcelino** [comment]
@kirkaleze, check balderdashy/waterline#705 for the latest on through associations. TL;DR: currently it's a bit of hit and miss.

**Comment by sailsbot** [comment]
Thanks for posting, @devinivy.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by dmarcelino** [comment]
Reopening as this was already picked for `0.11`.

**Comment by sailsbot** [comment]
Thanks for posting, @devinivy.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by particlebanana** [comment]
Tracked in the roadmap and it points back here so when @sailsbot eventually closes this it will still be around.

**Comment by sailsbot** [comment]
Thanks for posting, @devinivy.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by argordmel** [comment]
Hi Guys!

I've the same error. I've this models

Partner.js

```
        products: {
            collection: 'product',
            via: 'product',
            through: 'Partner_product'
        },
```

Product.js

```
        partners: {
            collection: 'partner',
            via: 'partner',
            through: 'Partner_product'
        },
```

Partner_product.js

```
        product: {
            index: true,
            model: 'product'
        },
        partner: {
            index: true,
            model: 'partner'
        },
```
I get the following error in the console:

```
this.collections[linkedCollection].junctionTable = true;
                                                 ^
TypeError: Cannot set property 'junctionTable' of undefined
```
But if I change the through attribute by 'partner_state' (all lowercase) it works, so can't use the model Partner_state.find() but partner_state.find() it works

Regards!
